File: apps/sandbox-runtime/internal/egress/route_resolver.go
package egress

import (
	"fmt"
	"net"
	"strings"

	"github.com/mistlehq/mistle/apps/sandbox-runtime/internal/startup"
)

type Resolver struct {
	routes []startup.EgressCredentialRoute
}

type NewResolverInput struct {
	RuntimePlan startup.RuntimePlan
}

func NewResolver(input NewResolverInput) Resolver {
	return Resolver{
		routes: input.RuntimePlan.EgressRoutes,
	}
}

type ResolveMatchingRouteInput struct {
	Host       string
	Method     string
	TargetPath string
}

func normalizeTargetPath(targetPath string) string {
	if targetPath == "" {
		return "/"
	}
	if strings.HasPrefix(targetPath, "/") {
		return targetPath
	}
	return "/" + targetPath
}

func containsStringIgnoreCase(values []string, target string) bool {
	for _, value := range values {
		if strings.EqualFold(value, target) {
			return true
		}
	}
	return false
}

func pathMatchesPrefixes(pathPrefixes []string, targetPath string) bool {
	for _, pathPrefix := range pathPrefixes {
		if strings.HasPrefix(targetPath, pathPrefix) {
			return true
		}
	}
	return false
}

func normalizeTargetHost(targetHost string) string {
	trimmedTargetHost := strings.TrimSpace(targetHost)
	if trimmedTargetHost == "" {
		return ""
	}

	host, _, err := net.SplitHostPort(trimmedTargetHost)
	if err == nil {
		return strings.ToLower(host)
	}

	return strings.ToLower(trimmedTargetHost)
}

func hostMatches(route startup.EgressCredentialRoute, targetHost string) bool {
	normalizedTargetHost := normalizeTargetHost(targetHost)
	for _, host := range route.Match.Hosts {
		if strings.EqualFold(host, normalizedTargetHost) {
			return true
		}
	}

	return false
}

func (resolver Resolver) ResolveMatchingRoute(input ResolveMatchingRouteInput) (startup.EgressCredentialRoute, bool, error) {
	normalizedTargetPath := normalizeTargetPath(input.TargetPath)

	matchingRoutes := make([]startup.EgressCredentialRoute, 0, 1)
	for _, route := range resolver.routes {
		if !hostMatches(route, input.Host) {
			continue
		}
		if len(route.Match.Methods) > 0 && !containsStringIgnoreCase(route.Match.Methods, input.Method) {
			continue
		}
		if len(route.Match.PathPrefixes) > 0 && !pathMatchesPrefixes(route.Match.PathPrefixes, normalizedTargetPath) {
			continue
		}

		matchingRoutes = append(matchingRoutes, route)
	}

	if len(matchingRoutes) == 0 {
		return startup.EgressCredentialRoute{}, false, nil
	}
	if len(matchingRoutes) > 1 {
		return startup.EgressCredentialRoute{}, false, fmt.Errorf(
			"multiple egress routes matched host=%q method=%q path=%q",
			normalizeTargetHost(input.Host),
			input.Method,
			normalizedTargetPath,
		)
	}

	return matchingRoutes[0], true, nil
}

File: apps/sandbox-runtime/src/runtime/proxy/route-resolver.ts
import type { EgressCredentialRoute } from "@mistle/integrations-core";

function normalizeTargetPath(targetPath: string): string {
  if (targetPath.length === 0) {
    return "/";
  }

  return targetPath.startsWith("/") ? targetPath : `/${targetPath}`;
}

function normalizePathPrefix(pathPrefix: string): string {
  const normalizedPathPrefix = normalizeTargetPath(pathPrefix);
  if (normalizedPathPrefix === "/") {
    return normalizedPathPrefix;
  }

  return normalizedPathPrefix.endsWith("/")
    ? normalizedPathPrefix.slice(0, -1)
    : normalizedPathPrefix;
}

function normalizeTargetHost(targetHost: string): string {
  const trimmedTargetHost = targetHost.trim().toLowerCase();
  if (trimmedTargetHost.length === 0) {
    return "";
  }

  if (trimmedTargetHost.startsWith("[")) {
    const endBracketIndex = trimmedTargetHost.indexOf("]");
    if (endBracketIndex >= 0) {
      return trimmedTargetHost.slice(1, endBracketIndex);
    }
  }

  const separatorIndex = trimmedTargetHost.lastIndexOf(":");
  if (separatorIndex > 0) {
    const host = trimmedTargetHost.slice(0, separatorIndex);
    const port = trimmedTargetHost.slice(separatorIndex + 1);
    if (port.length > 0 && Number.isInteger(Number(port))) {
      return host;
    }
  }

  return trimmedTargetHost;
}

function containsStringIgnoreCase(values: ReadonlyArray<string>, target: string): boolean {
  const normalizedTarget = target.toLowerCase();
  return values.some((value) => value.toLowerCase() === normalizedTarget);
}

function hostMatches(route: EgressCredentialRoute, targetHost: string): boolean {
  const normalizedTargetHost = normalizeTargetHost(targetHost);
  return route.match.hosts.some((host) => host.toLowerCase() === normalizedTargetHost);
}

function pathMatchesPrefixes(pathPrefixes: ReadonlyArray<string>, targetPath: string): boolean {
  return pathPrefixes.some((pathPrefix) => {
    const normalizedPathPrefix = normalizePathPrefix(pathPrefix);
    return (
      normalizedPathPrefix === "/" ||
      targetPath === normalizedPathPrefix ||
      targetPath.startsWith(`${normalizedPathPrefix}/`)
    );
  });
}

export type ResolveMatchingRouteInput = {
  routes: ReadonlyArray<EgressCredentialRoute>;
  host: string;
  method: string;
  targetPath: string;
};

export function resolveMatchingEgressRoute(
  input: ResolveMatchingRouteInput,
): EgressCredentialRoute | undefined {
  const normalizedTargetPath = normalizeTargetPath(input.targetPath);

  const matchingRoutes = input.routes.filter((route) => {
    if (!hostMatches(route, input.host)) {
      return false;
    }
    if (
      route.match.methods !== undefined &&
      route.match.methods.length > 0 &&
      !containsStringIgnoreCase(route.match.methods, input.method)
    ) {
      return false;
    }
    if (
      route.match.pathPrefixes !== undefined &&
      route.match.pathPrefixes.length > 0 &&
      !pathMatchesPrefixes(route.match.pathPrefixes, normalizedTargetPath)
    ) {
      return false;
    }

    return true;
  });

  if (matchingRoutes.length === 0) {
    return undefined;
  }

  if (matchingRoutes.length > 1) {
    throw new Error(
      `multiple egress routes matched host="${normalizeTargetHost(input.host)}" method="${input.method}" path="${normalizedTargetPath}"`,
    );
  }

  return matchingRoutes[0];
}
